packages/in_app_purchase/in_app_purchase_android/lib/src/billing_client_wrappers/billing_client_manager.dart

```diff
@@ -0,0 +1,155 @@
+    _connect();
+  }
+
+  /// Stream of `onPurchasesUpdated` events from the [BillingClient].
+  ///
+  /// This is a broadcast stream, so it can be listened to multiple times.
+  /// A "done" event will be sent after [dispose] is called.
+  late final Stream<PurchasesResultWrapper> purchasesUpdatedStream =
+      _purchasesUpdatedController.stream;
+
+  /// [BillingClient] instance managed by this [BillingClientManager].
+  ///
+  /// In order to access the [BillingClient], consider using [runWithClient]
+  /// and [runWithClientNonRetryable]
+  /// methods.
+  @visibleForTesting
+  late final BillingClient client = BillingClient(_onPurchasesUpdated);
+
+  final StreamController<PurchasesResultWrapper> _purchasesUpdatedController =
+  StreamController<PurchasesResultWrapper>.broadcast();
+
+  bool _isConnecting = false;
+  bool _isDisposed = false;
+
+  // Initialized immediately in the constructor, so it's always safe to access.
+  late Future<void> _readyFuture;
+
+  /// Executes the given [action] with access to the underlying [BillingClient].
+  ///
+  /// If necessary, waits for the underlying [BillingClient] to connect.
+  /// If given [action] returns [BillingResponse.serviceDisconnected], it will
+  /// be transparently retried after the connection is restored. Because
+  /// of this, [action] may be called multiple times.
+  ///
+  /// A response with [BillingResponse.serviceDisconnected] may be returned
+  /// in case of [dispose] being called during the operation.
+  ///
+  /// See [runWithClientNonRetryable] for operations that do not return
+  /// a subclass of [HasBillingResponse].
+  Future<R> runWithClient<R extends HasBillingResponse>(
+      Future<R> Function(BillingClient client) action,
+      ) async {
+    _debugAssertNotDisposed();
+    await _readyFuture;
+    final R result = await action(client);
+    if (result.responseCode == BillingResponse.serviceDisconnected &&
+        !_isDisposed) {
+      await _connect();
+      return runWithClient(action);
+    } else {
+      return result;
+    }
+  }
+
+  /// Executes the given [action] with access to the underlying [BillingClient].
+  ///
+  /// If necessary, waits for the underlying [BillingClient] to connect.
+  /// Designed only for operations that do not return a subclass
+  /// of [HasBillingResponse] (e.g. [BillingClient.isReady],
+  /// [BillingClient.isFeatureSupported]).
+  ///
+  /// See [runWithClient] for operations that return a subclass
+  /// of [HasBillingResponse].
+  Future<R> runWithClientNonRetryable<R>(
+      Future<R> Function(BillingClient client) action,
+      ) async {
+    _debugAssertNotDisposed();
+    await _readyFuture;
+    return action(client);
+  }
+
+  /// Ends connection to the [BillingClient].
+  ///
+  /// Consider calling [dispose] after you no longer need the [BillingClient]
+  /// API to free up the resources.
+  ///
+  /// After calling [dispose] :
+  /// - Further connection attempts will not be made;
```

Comment:
nit: To be consistent
```suggestion
  /// - Further connection attempts will not be made.
```